kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/data/ColumnVectorConverter.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal.data;
+
+import java.math.BigDecimal;
+import java.util.Optional;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.types.*;
+import io.delta.kernel.defaults.internal.data.vector.*;
+
+public class ColumnVectorConverter {
+    public static ColumnVector convertToOutputType(ColumnVector c, DataType outputType) {
+        DataType inputType = c.getDataType();
+        if (inputType == outputType) {
+            return c;
+        }
+        if (outputType instanceof BooleanType) {
+            // TODO: how to convert string to boolean?
+            // "true" "TRUE", "T" -> true ?
+            throw new IllegalStateException("convert to boolean type is not yet supported");
+        } else if (outputType instanceof ShortType) {
+            return convertToShort(c);
+        }  else if (outputType instanceof IntegerType) {
+            return convertToInteger(c);
+        } else if (outputType instanceof LongType) {
+            return convertToLong(c);
+        } else if (outputType instanceof DecimalType) {
+            return convertToDecimal(c);
+        } else if (outputType instanceof FloatType) {
+            return convertToFloat(c);
+        } else if (outputType instanceof DoubleType) {
+            return convertToDouble(c);
+        } else {
+            String msg = String.format(
+                    "`%s` is not supported to convert to data type %s`",
+                    inputType, outputType);
+            throw new IllegalStateException(msg);
+        }
+    }
+
+    protected static ColumnVector convertToShort(ColumnVector c) {
```

Comment:
There is too much boilerplate code here, every conversion have to handle all possible least common type conversion with the same destination outputType (ie. double have to handle byte, short, int, long, decimal, float, string) , I am not confident that this is the best way to handle this, next step is to play around with using Object[] to see if we can decrease boilerplate code.